### PR TITLE
Retry consuming messages

### PR DESCRIPTION
**Pull request by cristianormd**

### CODE REVIEW

File: lib/kafka_message_bus/consumer.ex [hunk 1]
```diff
@@ -17,19 +18,43 @@ defmodule KafkaMessageBus.Consumer do
...
-
+    value = value_from_message(message)
     Logger.debug(fn ->
-      "Got message: #{message.topic}/#{message.partition} -> #{message.key}: #{data}"
+      "Got message: #{message.topic}/#{message.partition} -> #{message.key}: #{value}"
     end)
-    message_processor.process(data)
+    try do
+      message_processor.process(value, message.key)
+    catch
```
Comment (jeffhsta) [inline lib/kafka_message_bus/consumer.ex]: Is it `catch` or `rescue` the better way to get the error?

File: lib/kafka_message_bus/consumer.ex [hunk 2]
```diff
@@ -17,19 +18,43 @@ defmodule KafkaMessageBus.Consumer do
...
   end
 
   defp execute_message(_, _), do: :ok
 
+  def enqueue_message_retry(value, key, message_processor) when @retry_strategy == :exq do
+    Exq.enqueue(Exq, "consume_queue", KafkaMessageBus.Consumer, [value, key, message_processor])
+  end
+
+  def enqueue_message_retry(value, key, message_processor) do
+    Logger.warn(fn ->
```
Comment (jeffhsta) [inline lib/kafka_message_bus/consumer.ex]: I think we don't need use anonymous function for `Logger.warn`

File: lib/kafka_message_bus/consumer.ex [hunk 3]
```diff
@@ -17,19 +18,43 @@ defmodule KafkaMessageBus.Consumer do
...
+      "Will not retry message with key: #{key} and value: #{value}"
+    end)
+  end
+
+  #this is the function that is run by Exq when retrying the message processing
```
Comment (jeffhsta) [inline lib/kafka_message_bus/consumer.ex]: You can delete comment in the code

File: lib/kafka_message_bus/consumer.ex [hunk 4]
```diff
@@ -17,19 +18,43 @@ defmodule KafkaMessageBus.Consumer do
...
+  def perform(value, key, message_processor) do
```
Comment (jeffhsta) [inline lib/kafka_message_bus/consumer.ex]: This `perform` function looks to me that should be in somewhere else not in this module

File: lib/kafka_message_bus/consumer.ex [hunk 5]
```diff
@@ -17,19 +18,43 @@ defmodule KafkaMessageBus.Consumer do
...
+    Logger.warn(fn ->
+      "Retrying message with key: #{key} and value: #{value}"
+    end)
+    message_processor.process(value, key)
+  end
+
+  def value_from_message(message) do
```
Comment (jeffhsta) [inline lib/kafka_message_bus/consumer.ex]: This function should be private

File: lib/kafka_message_bus/consumer.ex [hunk 6]
```diff
@@ -17,19 +18,43 @@ defmodule KafkaMessageBus.Consumer do
...
+    try do
+      message_processor.process(value, message.key)
+    catch
+      _ -> enqueue_message_retry(value, message.key, message_processor)
+    end
   end
 
   defp execute_message(_, _), do: :ok
 
+  def enqueue_message_retry(value, key, message_processor) when @retry_strategy == :exq do
```
Comment (jeffhsta) [inline lib/kafka_message_bus/consumer.ex]: This function should be private

File: lib/kafka_message_bus/consumer.ex [hunk 7]
```diff
@@ -1,6 +1,7 @@
 defmodule KafkaMessageBus.Consumer do
   require Logger
   @topics_and_processors Application.get_env(:kafka_message_bus, :consumers)
+  @retry_strategy Application.get_env(:kafka_message_bus, :retry_strategy)
```
Comment (goalves) [inline lib/kafka_message_bus/consumer.ex]: There is no default config for this.

File: lib/kafka_message_bus/consumer.ex (hunk 1 above)
Comment (cristianormd) [reply lib/kafka_message_bus/consumer.ex]: Catch. Rescue throws it back up.

File: lib/kafka_message_bus/consumer.ex (hunk 3 above)
Comment (cristianormd) [reply lib/kafka_message_bus/consumer.ex]: This comment is necessary.

File: lib/kafka_message_bus/consumer.ex (hunk 4 above)
Comment (cristianormd) [reply lib/kafka_message_bus/consumer.ex]: It is the retry mechanism for the consumer, I don't think it makes sense to separate this from the consumer itself.

File: lib/kafka_message_bus/consumer.ex (hunk 3 above)
Comment (jeffhsta) [reply lib/kafka_message_bus/consumer.ex]: If this comment is necessary it means your code is not clean enough

**Pull request merged**